sentry_sdk/spotlight.py

```diff
@@ -52,50 +61,137 @@ def capture_envelope(self, envelope):
 
+            self.sentry_sdk = sentry_sdk.api
+
             spotlight_client = sentry_sdk.api.get_client().spotlight
             if spotlight_client is None:
+                sentry_logger.warning(
+                    "Cannot find Spotlight client from SpotlightMiddleware, disabling the middleware."
+                )
                 return None
-
             # Spotlight URL has a trailing `/stream` part at the end so split it off
-            spotlight_url = spotlight_client.url.rsplit("/", 1)[0]
+            self._spotlight_url = urllib.parse.urljoin(spotlight_client.url, "../")
+
+        @property
+        def spotlight_script(self):
+            # type: (Self) -> Optional[str]
+            if self.__spotlight_script is None:
+                try:
+                    spotlight_js_url = urllib.parse.urljoin(
+                        self._spotlight_url, SPOTLIGHT_JS_ENTRY_PATH
+                    )
+                    req = urllib.request.Request(
+                        spotlight_js_url,
+                        method="HEAD",
+                    )
+                    urllib.request.urlopen(req)
+                    self.__spotlight_script = SPOTLIGHT_JS_SNIPPET_PATTERN.format(
+                        spotlight_js_url
+                    )
+                except urllib.error.URLError as err:
+                    sentry_logger.debug(
+                        "Cannot get Spotlight JS to inject at %s. SpotlightMiddleware will not be very useful.",
+                        spotlight_js_url,
+                        exc_info=err,
+                    )
+
+            return self.__spotlight_script
+
+        def process_response(self, _request, response):
+            # type: (Self, HttpRequest, HttpResponse) -> Optional[HttpResponse]
+            content_type_header = tuple(
+                p.strip()
+                for p in response.headers.get("Content-Type", "").lower().split(";")
+            )
+            content_type = content_type_header[0]
+            if len(content_type_header) > 1 and content_type_header[1].startswith(
+                CHARSET_PREFIX
+            ):
+                encoding = content_type_header[1][len(CHARSET_PREFIX) :]
+            else:
+                encoding = "utf-8"
+
+            if (
+                self.spotlight_script is not None
+                and not response.streaming
+                and content_type == "text/html"
+            ):
+                content_length = len(response.content)
+                injection = self.spotlight_script.encode(encoding)
+                injection_site = next(
+                    (
+                        idx
+                        for idx in (
+                            response.content.rfind(body_variant.encode(encoding))
+                            for body_variant in BODY_CLOSE_TAG_POSSIBILITIES
+                        )
+                        if idx > -1
+                    ),
+                    content_length,
+                )
```

Comment:
Can this be done with a regex? We could also get rid of `BODY_CLOSE_TAG_POSSIBILITIES` that way

Reply:
We probably can but it has a lot of drawbacks:

1. Would probably much slower
2. Would probably use way more memory, especially for large payloads
3. No way to support streaming responses (we don't support them right now but we _can_ if we try)